x-pack/heartbeat/security_linux.go

```diff
@@ -0,0 +1,222 @@
+
+package main
+
+import (
+	"fmt"
+	"os"
+	"os/user"
+	"runtime"
+	"strconv"
+	"syscall"
+
+	"github.com/elastic/beats/v7/libbeat/common/seccomp"
+	"kernel.org/pub/linux/libs/security/libcap/cap"
+)
+
+func init() {
+	// Here we set a bunch of linux specific security stuff.
+	// In the context of a container, where users frequently run as root, we follow BEAT_SETUID_AS to setuid/gid
+	// and add capabilities to make this actually run as a regular user. This also helps node in synthetics, which
+	// does not want to run as root. It's also just generally more secure.
+	if localUserName := os.Getenv("BEAT_SETUID_AS"); localUserName != "" && syscall.Geteuid() == 0 {
+		err := changeUser(localUserName)
+		if err != nil {
+			panic(err)
+		}
+		err = setCapabilities()
+		if err != nil {
+			panic(err)
+		}
+	}
+
+	switch runtime.GOARCH {
+	case "amd64", "386":
+		err := setSeccompRules()
+		if err != nil {
+			panic(err)
+		}
+	}
+}
+
+func changeUser(localUserName string) error {
+	localUser, err := user.Lookup(localUserName)
+	if err != nil {
+		return fmt.Errorf("could not lookup '%s': %s", localUser, err)
+	}
+	localUserUid, err := strconv.Atoi(localUser.Uid)
+	if err != nil {
+		return fmt.Errorf("could not parse UID '%s' as int: %s", localUser.Uid, err)
+	}
+	localUserGid, err := strconv.Atoi(localUser.Gid)
+	if err != nil {
+		return fmt.Errorf("could not parse GID '%s' as int: %s", localUser.Uid, err)
+	}
+	// We include the root group because the docker image contains many directories (data,logs)
+	// that are owned by root:root with 0775 perms. The heartbeat user is in both groups
+	// in the container, but we need to repeat that here.
+	err = syscall.Setgroups([]int{localUserUid, 0})
+	if err != nil {
+		return fmt.Errorf("could not prsetgroups: %s", err)
+	}
+
+	// Set the main group as 1000 so new files created are owned by the user's group
+	err = syscall.Setgid(1000)
+	if err != nil {
+		return fmt.Errorf("could not set gid to %d: %s", localUserGid, err)
+	}
+
+	// Note this is not the regular SetUID! Look at the 'cap' package docs for it, it preserves
+	// capabilities post-SetUID, which we use to lock things down immediately
+	err = cap.SetUID(localUserUid)
+	if err != nil {
+		return fmt.Errorf("could not setuid to %d: %s", localUserUid, err)
+	}
+
+	u, err := user.Lookup(localUserName)
+	if err != nil {
+		return fmt.Errorf("could not lookup local username: %s", err)
+	}
+
+	// This may not be necessary, but is good hygeine, we do some shelling out to node/npm etc.
+	// and $HOME should reflect the user's preferences
+	return os.Setenv("HOME", u.HomeDir)
+}
+
+func setCapabilities() error {
+	// Start with an empty capability set
+	newcaps := cap.NewSet()
+	// Both permitted and effective are required! Permitted makes the permmission
+	// possible to get, effective makes it 'active'
+	err := newcaps.SetFlag(cap.Permitted, true, cap.NET_RAW)
+	if err != nil {
+		return fmt.Errorf("error setting permitted setcap: %s", err)
```

Comment:
I recommend using `%w` rather than `%s` in all these `fmt.Errorf`s to wrap the original error.

Reply:
Ah, good catch this was originally a sprintf and panic